scylla/src/transport/errors.rs

```diff
@@ -871,32 +871,52 @@ pub enum CqlEventHandlingError {
 }
 
 /// An error type that occurred when executing one of:
-/// - QUERY
-/// - PREPARE
-/// - EXECUTE
-/// - BATCH
+/// - `QUERY`
+/// - `PREPARE`
+/// - `EXECUTE`
+/// - `BATCH`
 ///
-/// requests.
+/// requests. This error can appear during a single try
+/// of request execution. The retry decision is made based
+/// on this error.
 #[derive(Error, Debug)]
-pub(crate) enum UserRequestError {
+#[non_exhaustive]
+pub enum UserRequestError {
+    /// Failed to serialize CQL request.
     #[error("Failed to serialize CQL request: {0}")]
     CqlRequestSerialization(#[from] CqlRequestSerializationError),
+
+    /// Database sent a response containing some error with a message
     #[error("Database returned an error: {0}, Error message: {1}")]
     DbError(DbError, String),
+
+    /// Received a RESULT server response, but failed to deserialize it.
     #[error(transparent)]
     CqlResultParseError(#[from] CqlResultParseError),
+
+    /// Received an ERROR server response, but failed to deserialize it.
     #[error("Failed to deserialize ERROR response: {0}")]
     CqlErrorParseError(#[from] CqlErrorParseError),
+
+    /// Received an unexpected response from the server.
     #[error(
         "Received unexpected response from the server: {0}. Expected RESULT or ERROR response."
     )]
     UnexpectedResponse(CqlResponseKind),
+
+    /// A connection has been broken during query execution.
     #[error(transparent)]
     BrokenConnectionError(#[from] BrokenConnectionError),
+
+    /// Failed to deserialize frame body extensions.
     #[error(transparent)]
     BodyExtensionsParseError(#[from] FrameBodyExtensionsParseError),
+
+    /// Driver was unable to allocate a stream id to execute a query on.
     #[error("Unable to allocate stream id")]
     UnableToAllocStreamId,
+
+    /// Prepared statement id changed after repreparation.
```

Comment:
:camping:  Could we reorder variants of this enum so that they are grouped and sorted according to the moment in the request execution they can arise at?
I.e., serialization errors would come before deserialization errors, and `BodyExtensionsParseError` would go before `DbError` and `CqlResultParseError`.

This way it is easier to follow the possible errors along the request execution path.